src/cephadm/cephadm

```diff
@@ -5212,6 +5222,185 @@ def command_gather_facts():
 
 
+class CephadmDaemonHandler(BaseHTTPRequestHandler):
+    def do_GET(self):
+        # return the current state of the cache
+        self.send_response(200)
+        self.send_header('Content-type','application/json')
+        self.end_headers()
+        self.wfile.write(bytes(json.dumps(cephadm_cache), 'ascii'))
+    
+    def log_message(self, format, *args):
+        rqst = " ".join(args)
+        logger.info(f"client:{self.address_string()} [{self.log_date_time_string()}] {rqst}")
+
+
+class CephadmHTTPServer(ThreadingMixIn, HTTPServer):
+    allow_reuse_address = True
+    daemon_threads = True
+
+
+def cephadmd_install():
+    # open firewall port
+    # copy self to /usr/bin somewhere
+    # create the systemd unit file
+    # enable the systemd unit file
+    # start the systemd unit
+    pass
+
+def cephadmd_uninstall():
+    # shutdown the systemd unit
+    # disable it and remove it
+    # close the firewall port
+    pass
+
+def cephadmd_upgrade():
+    # copy self to the /usr/bin
+    # restart the systemd unit
+    pass
+
+
+class CephadmDaemon():
+    def __init__(self, fsid, port=None):
+        self.fsid = fsid
+        self.port = port
+        self.workers = []
+        self.http_server = None
+        self.stop = False
+
+    @property
+    def port_active(self):
+        return port_in_use(self.port)
+
+    @property
+    def can_run(self):
+        return not self.port_active
+    
+    def _scrape_host_facts(self, refresh_interval=10):
+        loop_delay = 0.1
+        ctr = 0
+        while True:
+            
+            if self.stop:
+                break
+
+            if ctr >= refresh_interval:
+                ctr = 0
+                logger.debug("executing host_facts scrape")
+                s_time = time.time()
+                facts = HostFacts()
+                elapsed = time.time() - s_time
+                data = json.loads(facts.dump())
+                with cephadm_cache_lock:
+                    cephadm_cache['host'] = {
+                        "scrape_timestamp": s_time,
+                        "scrape_duration_secs": elapsed,
+                        "data": data,                    
+                    }
+                logger.debug(f"completed host-facts scrape - {elapsed}s")
+            
+            time.sleep(loop_delay)
+            ctr += loop_delay
+        logger.info("host-facts thread stopped")
+
+    def _scrape_list_daemons(self, refresh_interval=20):
+        loop_delay = 0.1
+        ctr = 0
+        while True:
+            if self.stop:
+                break
+            
+            if ctr >= refresh_interval:
+                ctr = 0
+                logger.debug("executing list_daemons scrape")
+                s_time = time.time()
+                ld = list_daemons()
+                elapsed = time.time() - s_time
+                with cephadm_cache_lock:
+                    cephadm_cache['daemons'] = {
+                        "scrape_timestamp": s_time,
```

Comment:
how can we detect clock skew?

Reply:
To clarify - we have chrony as a pre-req to keep the host clocks in the cluster sync'd for ceph - so why is the above not acceptable?

Reply:
Take this more as a thought, instead of a requirement, please.

I've seen cases where the deployment was faster than ntp had the time to sync clocks. If we're going to deploy this as the first daemon on a host, we're a bit more exposed to potential clock skews than usual.

Reply:
ACK - so we're talking about the window between a scrape and the ntp sync - the implication being that the scrape time may not be 100%. Why not look at this differently : when we add a host we check for the time sync service, but don't check tracking - that to me is the place for this i.e. only accept the host once it has an enabled service, and it's in sync.